### PR TITLE
enhancements and fixes for FSDP and DeepSpeed

### LINKED ISSUE #285: Issues with saving model/optimizer and loading them back
Hello @sgugger ,

Came across multiple related issues regarding this - https://github.com/huggingface/accelerate/issues/242, https://github.com/huggingface/accelerate/issues/154 . They were all closed with this PR - https://github.com/huggingface/accelerate/pull/255, but unfortunately the PR doesn't seem to have much documentation. 

I was looking for specifically: saving a model, it's optimizer state, LR scheduler state, it's random seeds/states, epoch/step count, and other related similar states for reproducible training runs and resuming them correctly. 

I know there's this very brief doc here: [here](https://huggingface.co/docs/accelerate/main/en/accelerator#accelerate.Accelerator.save_state) and [here](https://huggingface.co/docs/accelerate/main/en/accelerator#accelerate.Accelerator.load_state) , but it looks like there are still few grey areas not documented currently regarding it's usage. 
a) My question is specifically that, like in the official example here: [link](https://github.com/huggingface/transformers/blob/master/examples/pytorch/text-classification/run_glue_no_trainer.py#L455) that saves using `save_pretrained` only in the main process, should I be using these only in the main process (both save/load) too, and in case of load_state I will have to call prepare() after load_state is done to prepare them for multi-gpu training/inference after that is done (or does load_state do all of that internally itself?)? 
b) Does the save_state method call `save_pretrained` methods for the model internally or do I have to do both? FWIW, I'm using HF's BERT and other pretrained models from the transformers lib, so if there are any other specialized methods specifically for those then please advise on the same. If there's any simple toy example that already uses these new checkpointing methods, and if you can help share that'd be pretty helpful! 

The last release seems to be way back in Sept 2021 - https://github.com/huggingface/accelerate/releases/tag/v0.5.1 - and the PR is just about a month old. Any plans for a soonish version-bump release of accelerate?

Request: If some more detailed examples can be added to the docs that'd be really awesome and help clarify about some of these specifics to users more easily! 

Thanks so much in advance! :)

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?

1. Enables `StateDictType` functionality of FSDP for checkpointing `Full` and `Sharded` model and optimizer states.
2. Enabled DeepSpeed checkpointing functionality to save and load `Full` and `Sharded` model, optimizer and scheduler states.
3. Fixes #285 related to deepspeed and the optimizer issue of #489 when using deepspeed
4. Renaming all FSDP arguments with `fsdp` prefix for clarity and readability
5. Fixing `accelerator.wait_for_everyone` to account for FSDP

One can now use `accelerator.save_state(output_dir)` and `accelerator.load_state(input_dir)` without having to deal with complexities of DeepSpeed and FSDP.

#### ToDo
- [x] Write tests
- [x]  Run experiments to verify
- [x] update the docs related to this PR

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! I left a few comments, but mainly take care of not breaking backward compatibility by changing public functions.

File: src/accelerate/commands/config/cluster.py [hunk 1]
```diff
@@ -250,6 +251,15 @@ def get_cluster_input():
...
             )
+            fsdp_state_dict_type_query = "What should be your FSDP's state dict type ("
+            for i, state_dict_type in enumerate(FSDP_STATE_DICT_TYPE):
+                fsdp_state_dict_type_query += f"[{i}] {state_dict_type}, "
+            fsdp_state_dict_type_query = fsdp_state_dict_type_query[:-2] + ")? [0]: "
+            fsdp_config["fsdp_state_dict_type"] = _ask_field(
+                fsdp_state_dict_type_query,
+                lambda x: FSDP_STATE_DICT_TYPE[int(x)],
+                default=FSDP_STATE_DICT_TYPE[0],
+            )
```
[the remark this answers is not in the document]
Comment (sgugger) [reply src/accelerate/commands/config/cluster.py:262]: It would be more readable to just have a value.

File: src/accelerate/utils/dataclasses.py [hunk 2]
```diff
@@ -472,6 +492,17 @@ def __post_init__(self):
             if prefetch_policy != FSDP_BACKWARD_PREFETCH[-1]:
                 self.backward_prefetch = BackwardPrefetch(FSDP_BACKWARD_PREFETCH.index(prefetch_policy) + 1)
 
+        if self.state_dict_type is None:
+            state_dict_type_policy = os.environ.get("FSDP_STATE_DICT_TYPE", FSDP_STATE_DICT_TYPE[0])
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/utils/dataclasses.py:496]: As mentioned above, we are separately maintaining it in our repo and hence this should not be an issue.

File: src/accelerate/commands/config/cluster.py [hunk 3]
```diff
@@ -210,12 +211,12 @@ def get_cluster_input():
             for i, strategy in enumerate(FSDP_SHARDING_STRATEGY):
                 sharding_strategy_query += f"[{i+1}] {strategy}, "
             sharding_strategy_query = sharding_strategy_query[:-2] + ")? [1]: "
-            fsdp_config["sharding_strategy"] = _ask_field(
+            fsdp_config["fsdp_sharding_strategy"] = _ask_field(
```
[the remark this answers is not in the document]
Comment (sgugger) [reply src/accelerate/commands/config/cluster.py:214]: Changing the launcher is fair game, but existing configs breaking when updating Accelerate is a bit too annoying to the user I think.

File: src/accelerate/commands/launch.py [hunk 4]
```diff
@@ -141,19 +141,19 @@ def launch_command_parser(subparsers=None):
         help="Whether to use fsdp.",
     )
     parser.add_argument(
-        "--offload_params",
+        "--fsdp_offload_params",
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/commands/launch.py:144]: As mentioned above, I thought this was fair game as we specified FSDP integration was experimental given that FSDP itself has changed a lot. This was more around readability/grouping for folks to easily understand all the variables necessary for FSDP as there are a lot more now. If you think we can do away with it at the expense of readability/clarity, I will revert the variable name changes.

File: src/accelerate/commands/config/cluster.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/commands/config/cluster.py:262]: You mean for the `default` or entire query?

File: src/accelerate/utils/dataclasses.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply src/accelerate/utils/dataclasses.py:496]: But it's not readable.

[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/utils/dataclasses.py]: Done.

File: src/accelerate/commands/launch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/commands/launch.py:144]: Done. Added deprecation args with the warnings.

File: src/accelerate/commands/config/cluster.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/commands/config/cluster.py:214]: Understood. Added code to handle this wherein old configs should work without any issues.

File: src/accelerate/checkpointing.py [hunk 5]
```diff
@@ -109,11 +133,21 @@ def save_accelerator_state(
...
-def load_accelerator_state(input_dir, models, optimizers, schedulers, process_index, scaler=None):
+def load_accelerator_state(
+    accelerator,
+    input_dir: str,
+    models: list,
+    optimizers: list,
+    schedulers: list,
+    process_index: int,
+    scaler: GradScaler = None,
+):
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/checkpointing.py]: Done.

File: src/accelerate/checkpointing.py [hunk 6]
```diff
@@ -68,20 +72,40 @@ def save_accelerator_state(
...
+            logger.info("Saving DeepSpeed Model and Optimizer")
+            ckpt_id = f"{MODEL_NAME}" if i == 0 else f"{MODEL_NAME}_{i}"
+            model.save_checkpoint(output_dir, ckpt_id)
+            logger.info(f"DeepSpeed Model and Optimizer saved to output dir {os.path.join(output_dir, ckpt_id)}")
+        else:
+            state = accelerator.get_state_dict(model, unwrap=False)
+            weights_name = f"{MODEL_NAME}.bin" if i == 0 else f"{MODEL_NAME}_{i}.bin"
+            output_model_file = os.path.join(output_dir, weights_name)
+            save(state, output_model_file)
+            logger.info(f"Model weights saved in {output_model_file}")
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/checkpointing.py]: Done.

File: src/accelerate/checkpointing.py [hunk 7]
```diff
@@ -43,8 +44,9 @@
 
 
 def save_accelerator_state(
+    accelerator,
     output_dir: str,
-    model_states: List[dict],
+    models: list,
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/checkpointing.py]: Done.

File: src/accelerate/commands/config/cluster.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply src/accelerate/commands/config/cluster.py:262]: For the default

**Review by sgugger — APPROVED** [review]
Thanks for iterating!

**Comment by pacman100** [comment]
`make test_deepspeed` results below (~6 minutes) :

Screenshot 2022-07-26 at 1 11 22 PM

`make test_fsdp` results below (~4 minutes):

Screenshot 2022-07-26 at 1 25 05 PM

**Review by sgugger — APPROVED** [review]
LGTM, thanks for working on this!
